### PR TITLE
fixing index clamping for upsample nearest kernel backward

### LINKED ISSUE #51036: Incorrect gradients for F.interpolate on GPU when mode=nearest
## 🐛 Bug

Function `F.interpolate` gives incorrect gradients on GPU when using `mode='nearest'`.

## To Reproduce

``` python
import torch
import torch.nn.functional as F


for seed in range(0, 20):
    torch.manual_seed(seed)

    in_shape = torch.randint(low=10, high=50, size=(2,)).tolist()
    in_map = torch.randn(2, 256, *in_shape, device='cuda', requires_grad=True)

    out_shape = torch.randint(low=10, high=50, size=(2,)).tolist()
    inter_map = F.interpolate(in_map, size=tuple(out_shape), mode='nearest')

    output_map = torch.mul(inter_map, 0.0)
    output_map.sum().backward()

    max_grad = in_map.grad.abs().max().item()
    print(f"({seed:02d}) Max grad: {max_grad}")
```

## Output on my GPU

```
(00) Max grad: 0.0
(01) Max grad: 0.0
(02) Max grad: 0.0
(03) Max grad: 0.0
(04) Max grad: 0.0
(05) Max grad: 0.0
(06) Max grad: 0.0
(07) Max grad: 0.0
(08) Max grad: 0.0
(09) Max grad: 0.0
(10) Max grad: 0.0
(11) Max grad: 0.0
(12) Max grad: 0.0
(13) Max grad: 0.0
(14) Max grad: 2.55876088142395
(15) Max grad: 0.0
(16) Max grad: 0.0
(17) Max grad: 0.0
(18) Max grad: 0.0
(19) Max grad: 0.0
```

## Expected behavior

All gradients should be zero.

## Environment

```
PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Fedora 33 (Thirty Three) (x86_64)
GCC version: (GCC) 10.2.1 20201125 (Red Hat 10.2.1-9)
Clang version: 11.0.0 (Fedora 11.0.0-2.fc33)
CMake version: version 3.18.4

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1050 Ti
Nvidia driver version: 460.32.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.7.0
[pip3] torchvision==0.8.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.2.89              hfd86e86_1  
[conda] libblas                   3.8.0                    21_mkl    conda-forge
[conda] libcblas                  3.8.0                    21_mkl    conda-forge
[conda] liblapack                 3.8.0                    21_mkl    conda-forge
[conda] liblapacke                3.8.0                    21_mkl    conda-forge
[conda] mkl                       2020.2                      256  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.2.0            py38h23d657b_0  
[conda] mkl_random                1.1.1            py38h0573a6f_0  
[conda] numpy                     1.19.2           py38h54aff64_0  
[conda] numpy-base                1.19.2           py38hfa32c7d_0  
[conda] pytorch                   1.7.0           py3.8_cuda10.2.89_cudnn7.6.5_0    pytorch
[conda] torchvision               0.8.1                py38_cu102    pytorch
```

## Additional context

- I've tested this on different GPU's. All of them gave errors, but not always the same error and not always for the same seed.
- On a same GPU, the same seed might not always give an error depending on the used seed `range` in the code sample.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @ngimel

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes #51036

### CODE REVIEW

**Comment by ngimel** [comment]
Can you please add a test? I'm pretty sure currently it should be failing, as it reads some random data.

**Comment by jjsjann123** [comment]
Will do. Apparently it's breaking existing tests so there's probably something wrong with indexing :)

File: aten/src/ATen/native/cuda/UpSample.cuh [hunk 1]
```diff
@@ -253,7 +253,7 @@ __device__ __forceinline__ static int nearest_neighbor_bw_compute_source_index(
     int dst_index,
     int output_size) {
   const int src_index =
-      static_cast<int>(ceilf(dst_index * scale));
+      min(static_cast<int>(ceilf(dst_index * scale)), output_size - 1);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSample.cuh:256]: given that subsequent loop is `for (int y = src_y; y < src_y_up; y++)`, I'd rather have `output_size` here, and no `+1` at the call site. Maximum returned index will be `output_size`, and it properly will never be accessed.

**Comment by jjsjann123** [comment]
Added test for illegal memory access.
I verified the test failure on build without the fix. @ngimel

**Review by ngimel — APPROVED** [review]
